src/pgduckdb_ruleutils.cpp

```diff
@@ -16,4 +28,215 @@ pgduckdb_function_name(Oid function_oid) {
 	auto func_name = get_func_name(function_oid);
 	return psprintf("system.main.%s", quote_identifier(func_name));
 }
+
+/*
+ * generate_relation_name computes the fully qualified name of the relation in
+ * DuckDB for the specified Postgres OID. This includes the DuckDB database name
+ * too.
+ */
+char *
+pgduckdb_relation_name(Oid relation_oid) {
+
+	HeapTuple tp = SearchSysCache1(RELOID, ObjectIdGetDatum(relation_oid));
+	if (!HeapTupleIsValid(tp))
+		elog(ERROR, "cache lookup failed for relation %u", relation_oid);
+	Form_pg_class relation = (Form_pg_class)GETSTRUCT(tp);
+	const char *relname = NameStr(relation->relname);
+	/*
+	 * XXX: Should this be using get_namespace_name instead, the difference is
+	 * that it will store temp tables in the pg_temp_123 style schemas instead
+	 * of plain pg_temp. I don't think this really matters.
+	 */
+	const char *nspname = get_namespace_name_or_temp(relation->relnamespace);
+	const char *dbname;
+
+	if (relation->relam == pgduckdb::DuckdbTableAmOid()) {
+		dbname = "memory";
+	} else {
+		dbname = "pgduckdb";
+	}
+
+	char *result = psprintf("%s.%s.%s", quote_identifier(dbname), quote_identifier(nspname), quote_identifier(relname));
+
+	ReleaseSysCache(tp);
+
+	return result;
+}
+
+/*
+ * pgduckdb_get_tabledef returns the definition of a given table. This
+ * definition includes table's schema, default column values, not null and check
+ * constraints. The definition does not include constraints that trigger index
+ * creations; specifically, unique and primary key constraints are excluded.
+ *
+ * TODO: Add support indexes, primary keys and unique constraints.
+ *
+ * This function is inspired by the pg_get_tableschemadef_string function in
+ * the following patch that I (Jelte) submitted to Postgres in 2023:
+ * https://www.postgresql.org/message-id/CAGECzQSqdDHO_s8=CPTb2+4eCLGUscdh=KjYGTunhvrwcC7ZSQ@mail.gmail.com
+ */
+char *
+pgduckdb_get_tabledef(Oid relation_oid) {
```

Comment:
I wonder if we could re-use the C++ test framework I started sketching to unit-test this method? (seems like a good candidate :-)

Reply:
I think it could be, but I think getting good test coverage here with the normal regression tests is actually pretty easy. And it's actually good that we have these end to end tests here. For now I'll leave this as is.